### PR TITLE
HIVE-23975: Reuse evicted keys from aggregation buffers

### PR DESCRIPTION
**Pull request by mustafaiman**
Change-Id: I17bf742f15de6f81144d485abf94c375d9a7e647

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByOperator.java [hunk 1]
```diff
@@ -514,7 +526,8 @@ private void prepareBatchAggregationBufferSets(VectorizedRowBatch batch) throws
           // is very important to clone the keywrapper, the one we have from our
           // keyWrappersBatch is going to be reset/reused on next batch.
           aggregationBuffer = allocateAggregationBuffer();
```
Comment (rbalamohan) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByOperator.java:528]: allocateAggregationBuffer can be reused with the flushed out entries.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/wrapper/VectorHashKeyWrapperGeneral.java [hunk 2]
```diff
@@ -262,6 +255,138 @@ private void duplicateTo(VectorHashKeyWrapperGeneral clone) {
...
+        System.arraycopy(byteStarts, 0, clone.byteStarts, 0, byteValues.length);
+        for (int i = 0; i < byteValues.length; ++i) {
+          // avoid allocation/copy of nulls, because it potentially expensive.
+          // branch instead.
+          if (byteLengths[i] != -1) {
+            if (clone.byteValues[i] != null && clone.byteValues[i].length >= byteValues[i].length) {
+              System.arraycopy(byteValues[i], byteStarts[i], clone.byteValues[i], 0, byteLengths[i]);
+            } else {
+              clone.byteValues[i] = Arrays.copyOfRange(byteValues[i],
+                  byteStarts[i], byteStarts[i] + byteLengths[i]);
```
Comment (rbalamohan) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/wrapper/VectorHashKeyWrapperGeneral.java:361]: Why is this needed? Wouldn't be it always from 0 - bytelength?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByOperator.java (hunk 1 above)
Comment (mustafaiman) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByOperator.java:528]: @rbalamohan I have another patch for that.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/wrapper/VectorHashKeyWrapperGeneral.java (hunk 2 above)
Comment (mustafaiman) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/wrapper/VectorHashKeyWrapperGeneral.java:361]: `clone.byteStarts[i]` is always zero but `byteStarts[i]` can be different values, depending on how it was assigned initially in `VectorHashKeyWrapperBatch#assignString` methods. Example: assignStringBullsNoRepeatingSelection
